compiler/rustc_resolve/src/late/diagnostics.rs

```diff
@@ -1289,25 +1289,43 @@ impl<'a: 'ast, 'ast, 'tcx> LateResolutionVisitor<'a, '_, 'ast, 'tcx> {
                 PathSource::Expr(_) | PathSource::TupleStruct(..) | PathSource::Pat => {
                     let span = find_span(&source, err);
                     err.span_label(self.r.def_span(def_id), &format!("`{path_str}` defined here"));
-                    let (tail, descr, applicability) = match source {
-                        PathSource::Pat | PathSource::TupleStruct(..) => {
-                            ("", "pattern", Applicability::MachineApplicable)
-                        }
-                        _ => (": val", "literal", Applicability::HasPlaceholders),
-                    };
 
+                    let (tail, descr, applicability, old_fields) = match source {
+                        PathSource::Pat => ("", "pattern", Applicability::MachineApplicable, None),
+                        PathSource::TupleStruct(_, args) => (
+                            "",
+                            "pattern",
+                            Applicability::MachineApplicable,
+                            Some(
+                                args.iter()
+                                    .map(|a| {
+                                        self.r.tcx.sess.source_map().span_to_snippet(*a).unwrap()
+                                    })
+                                    .collect::<Vec<String>>(),
+                            ),
```

Comment:
Last thing: can you remove the unwrap, falling back to the bad suggestion in case sources are not available?

Something like `args.iter().map(|&a| self.r.tcx.sess.source_map().span_to_snipper(a).ok()).collect::<Option<Vec<String>>>()` should work.